### PR TITLE
feat(collateral): harden dual-state collateral contract and validator integration

### PR DESCRIPTION
**Pull request by itzlambda**
## Summary
This PR hardens and completes Basilica’s collateral integration across:
1. the upgradeable collateral contract,
2. validator event ingestion and state persistence, and
3. collateral tooling used in localnet workflows.

Primary outcomes are stronger reclaim/slash invariants, correct dual-asset state handling (TAO + alpha), and end-to-end SHA-256 evidence consistency.

### 1) Contract hardening and lifecycle correctness (`crates/collateral-contract`)
- Strengthened core invariants in `CollateralUpgradeable` for deposit, reclaim, deny, and slash flows.
- Fixed edge cases around pending reclaims and slashes so collateral cannot be stranded by lifecycle races.
- Removed mutable coldkey setter behavior and moved to derived/immutable contract coldkey initialization.
- Added `taoDepositsEnabled` and `alphaDepositsEnabled` toggles.
- Added `minAlphaCollateralIncrease`.
- Standardized evidence fields/checksums to SHA-256 (`bytes32`) across contract events and calls.
- Aligned naming for clarity (`collaterals` -> `taoCollaterals`, `CONTRACT_HOTKEY` -> `validatorHotkey`).
- Removed stale/legacy contract artifacts and shifted upgrade testing to a dedicated upgrade mock flow.

### 2) Validator integration and state consistency (`crates/basilica-validator`)
- Expanded collateral event handling to cover full lifecycle events (`Deposit`, `ReclaimProcessStarted`, `Denied`, `Reclaimed`, `Slashed`).
- Updated persistence to track TAO and alpha separately, including pending reclaim state and reclaim request linkage.
- Made per-block collateral event application transactional to avoid partial state application.
- Updated slash execution path to use full SHA-256 checksum payloads and aligned slash fraction behavior with percentage semantics.
- Added optional collateral `rpc_url` override support for validator configuration.

### 3) CLI/bindings and API surface alignment (`collateral-cli` + Rust library)
- Updated Rust bindings and CLI command flow to match the new contract ABI and event model.
- Reclaim/deny/slash checksum inputs now use SHA-256 naming and types.
- Query/API naming is aligned to explicit TAO vs alpha collateral reads.
- Improved network/contract resolution through env-aware CLI defaults.

### 4) Localnet collateral workflow improvements (`scripts/*`)
- Added collateral setup/deploy/integration/e2e scripts for repeatable local testing.
- Wired collateral deployment into localnet startup flow and validator contract-address patching.
- Updated localnet bootstrap behavior to better support collateral lifecycle testing.

## Behavioral/API Changes to Validate
- Evidence checksum schema changed to SHA-256 (`bytes32`) end-to-end.
- Reclaim/slash behavior now enforces stricter pending-state consistency.
- ABI/event payloads changed; downstream clients should use updated bindings.
- Validator collateral config supports `rpc_url` override.

## Summary by CodeRabbit

* **New Features**
  * Full dual-state collateral UX: TAO + Alpha deposits, reclaim (deny/finalize), trustee slashing, and CLI collateral commands.
  * New collateral queries: all-collaterals, all-reclaims, active counts, and pagination.
  * Local tooling and tests: localnet setup, deploy/upgrade helpers, e2e validator tests, and utility scripts.

* **Improvements**
  * Alpha unit handling standardized to RAO; clearer token formatting in CLI outputs.
  * Checksum handling switched to SHA-256 and display improvements; expanded documentation and examples.

### CODE REVIEW

**Review by epappas — CHANGES_REQUESTED** [review]
# PR #372 Review: feat(collateral): harden dual-state collateral contract and validator integration

## Summary

This PR is a substantial change (+5904/-4305 across 51 files) that hardens Basilica's collateral integration across three layers: (1) the upgradeable Solidity collateral contract with dual-asset TAO + alpha support, (2) the Rust-side validator event ingestion and state persistence, and (3) localnet tooling and deployment scripts.

Three specialized review agents analyzed the PR in parallel: **blockchain-developer** (smart contract security and correctness), **rust-engineer** (Rust code quality and architecture), and **security-engineer** (cross-cutting security analysis and threat modeling).

---

## Issues Found

### CRITICAL Issues (Must Fix Before Merge)

#### C-1: Both `taoEnabled` and `alphaEnabled` can be set simultaneously, causing double-spending of `msg.value`

**Classification**: Smart Contract Fund Safety Vulnerability
**File**: `crates/collateral-contract/src/CollateralUpgradeable.sol`, lines 218-248 (setters) and 280-340 (deposit)

`setTaoEnabled` and `setAlphaEnabled` are independent admin toggles with no mutual exclusion. If an admin enables both:
1. `deposit()` at line 297 records `d.taoAmount += msg.value` (ETH stays in contract)
2. `deposit()` at line 318 calls `stakingPrecompile.addStake{value: msg.value}(hotkey, netuid)` (sends the **same** ETH to staking)

The same `msg.value` is counted as both TAO collateral AND alpha collateral, effectively draining the contract's ETH pool.

**Impact**: Contract insolvency. If Miner A deposits 1 ETH (TAO-only), then admin enables both modes, then Miner B deposits 1 ETH (both modes active) -- Miner B's `addStake` sends the ETH to staking, and the contract now has 0 ETH but tracks 2 ETH in TAO deposits.

**Proposed Fix**: Add mutual exclusion to the setter functions:
```solidity
function setTaoEnabled(bool _enabled) external onlyRole(DEFAULT_ADMIN_ROLE) {
    if (_enabled) require(!alphaEnabled, "Cannot enable TAO while alpha is enabled");
    taoEnabled = _enabled;
    emit TaoEnabledUpdated(_enabled);
}
```

---

#### C-2: `pendingReclaimCount` not decremented on full slash when reclaim was pending

**Classification**: State Inconsistency Bug
**File**: `crates/collateral-contract/src/CollateralUpgradeable.sol`, lines 740-755

When a node with a pending reclaim is fully slashed, the `_slash` function clears `d.reclaimRequest = ReclaimStatus.NONE` on line 749 but does NOT decrement `pendingReclaimCount[miner]`. Compare with `denyReclaim` (line 599) and `_finalizeReclaim` (line 523) which both correctly decrement.

**Impact**: `pendingReclaimCount` becomes permanently inflated. The `getPendingReclaimCount` view function returns incorrect data to off-chain systems (validator, CLI).

**Proposed Fix**:
```solidity
// In _slash, before clearing reclaimRequest:
if (d.reclaimRequest == ReclaimStatus.PENDING) {
    pendingReclaimCount[miner] -= 1;
}
d.reclaimRequest = ReclaimStatus.NONE;
```

---

#### C-3: `nodeToMiner` mapping not cleared on full slash, preventing node re-use

**Classification**: State Inconsistency / Denial of Service
**File**: `crates/collateral-contract/src/CollateralUpgradeable.sol`, lines 740-755

After a 100% slash, `taoAmount` and `alphaAmount` are zero, but `nodeToMiner[node]` retains the old miner address. The `deposit` function (line 269) checks `require(nodeToMiner[node] == address(0) || nodeToMiner[node] == msg.sender)`, so no other miner can deposit for that node. The only recovery is a manual `clearNodeMinerMapping` call by the trustee.

**Impact**: Slashed nodes become permanently unusable by other miners without manual intervention.

**Proposed Fix**:
```solidity
// In _slash, after zeroing amounts:
if (d.taoAmount == 0 && d.alphaAmount == 0) {
    delete nodeToMiner[node];
}
```

---

#### C-4: Collateral event scan loop has no graceful shutdown mechanism

**Classification**: Operational Reliability
**File**: `crates/basilica-validator/src/collateral/collateral_scan.rs`, lines 28-41

The `start()` method runs an infinite `loop` inside `tokio::select!` but has no `CancellationToken`. This means the event scan loop can never be gracefully stopped. The project's established pattern (used in rental monitoring and billing telemetry) uses `CancellationToken` for shutdown.

**Impact**: Process termination could interrupt a partially-committed block scan. The loop never exits cleanly.

**Proposed Fix**: Accept a `CancellationToken` and add a cancellation branch:
```rust
pub async fn start(&mut self, cancel: CancellationToken) -> Result<()> {
    loop {
        tokio::select! {
            _ = cancel.cancelled() => return Ok(()),
            _ = interval.tick() => { /* scan */ }
        }
    }
}
```

---

### HIGH Severity Issues (Strongly Advised to Fix Before Merge)

#### H-1: Trustee key compromise enables complete fund theft (no on-chain rate limiting)

**Classification**: Security Architecture
**File**: `crates/collateral-contract/src/CollateralUpgradeable.sol`, lines 508-556

`TRUSTEE_ROLE` has unrestricted ability to slash 100% of every miner's collateral (TAO + alpha) and direct it to the trustee address. The only safeguard is the off-chain `SlashRateLimiter` in the validator Rust code, which does not protect against direct contract interaction.

**Impact**: Total loss of all deposited collateral if trustee key is compromised.

**Recommendation**: Consider on-chain rate limiting (max slash per block/timeframe), multi-sig, or timelock for slashes above a threshold. At minimum, document key custody requirements.

---

#### H-2: `requestReclaim` clears `nodeToMiner` before finalization, enabling slash-evasion

**Classification**: Smart Contract Logic Vulnerability
**File**: `crates/collateral-contract/src/CollateralUpgradeable.sol`, line 435

`requestReclaim` deletes `nodeToMiner[node]` immediately upon request (before finalization). Another miner can immediately deposit for the same node. The `slash` function looks up the miner via `nodeToMiner`, so it would target the new miner instead of the misbehaving one. A malicious miner could request reclaim + have an accomplice deposit to shield from slashing.

**Impact**: Wrong miner gets slashed; enables slash evasion.

**Proposed Fix**: Keep `nodeToMiner` until reclaim is finalized, or add a node-level lock during pending reclaim.

---

#### H-3: `receive()` function allows arbitrary ETH deposits without tracking

**Classification**: Fund Stranding
**File**: `crates/collateral-contract/src/CollateralUpgradeable.sol`, line 947

The `receive() external payable {}` function accepts any ETH without accounting. While necessary for `removeStake` precompile returns, anyone can send ETH to the contract with no recovery mechanism.

**Proposed Fix**: Add an admin sweep function for excess ETH, or restrict `receive()` to only accept from the staking precompile.

---

#### H-4: No event emission for `setNodeMinerMapping` and `clearNodeMinerMapping`

**Classification**: Auditability Gap
**File**: `crates/collateral-contract/src/CollateralUpgradeable.sol`, lines 780-812

These functions modify the critical `nodeToMiner` mapping but do not emit events. All other state-mutating functions emit events. The validator's off-chain event scanner cannot track these changes.

---

#### H-5: `MAX_BLOCKS_PER_SCAN` defined but never enforced in scanning

**Classification**: Operational Reliability
**File**: `crates/collateral-contract/src/config.rs:21` and `crates/basilica-validator/src/collateral/collateral_scan.rs:57-58`

`MAX_BLOCKS_PER_SCAN = 1000` is defined but `scan_events()` queries from `from_block` to `current_block` in a single RPC call with no chunking. On first run (from block 0) or after extended downtime, this will timeout or be rejected by the RPC node.

**Proposed Fix**: Implement block range chunking in the scan loop.

---

#### H-6: `compute_slash_amount` has precision edge case -- small fractions cause 100% slash

**Classification**: Financial Calculation Bug
**File**: `crates/basilica-validator/src/collateral/slash_executor.rs`, lines 281-298

The slash amount computation rounds to integer percentage. A `slash_fraction` of `0.004` (0.4%) would compute `numerator = 0`, and the function returns full collateral (100% slash). For `slash_fraction = 0.10` with collateral = 1 RAO: `1 * 10 / 100 = 0`, so it returns 1 (100% slash instead of 10%).

**Proposed Fix**: Use basis points (10000) instead of percentage (100) for better precision.

---

### MEDIUM Severity Issues

#### M-1: Deposit after reclaim request inflates returned amount

**File**: `CollateralUpgradeable.sol`, lines 254-380

A miner can deposit more after requesting reclaim; `_finalizeReclaim` returns the FULL balance including post-request deposits (`d.taoAmount`). The trustee cannot predict the exact payout.

**Proposed Fix**: Block deposits while reclaim is pending, or snapshot amounts at request time.

---

#### M-2: Network string parsing duplicated with inconsistent fallback behavior

**Files**: `collateral_scan.rs:46-51` (falls through to Mainnet silently) vs `slash_executor.rs:596-610` (returns error)

A config typo could cause the scanner to connect to mainnet while the slash executor errors out. Extract a shared `parse_network()` function.

---

#### M-3: `get_collateral_amount()` returns TAO despite API name suggesting generic collateral

**File**: `collateral_persistence.rs:178-187`

Confusing API could lead callers to use TAO for eligibility checks when alpha should be used. The wrapper `get_tao_collateral_amount()` delegates to this same method, making one redundant.

---

#### M-4: `default_shadow_mode()` returns `false` -- live slashing by default

**File**: `crates/basilica-validator/src/config/collateral.rs:206-208`

A validator that enables collateral config without explicitly setting `shadow_mode = true` will immediately execute real on-chain slashes. Consider defaulting to `true` for safety.

---

#### M-5: ABI file naming inconsistency (`CollateralUpgradable` vs `CollateralUpgradeable`)

**File**: `crates/collateral-contract/src/CollateralUpgradableABI.json`

Missing 'e' in "Upgradeable". Should be `CollateralUpgradeableABI.json`.

---

#### M-6: RAO conversion precision loss can strand dust in staking

**File**: `CollateralUpgradeable.sol:855-858`

`_convertToRao` truncates via integer division by 1e9. Sub-gwei deposits lose dust permanently. Enforce `msg.value % RAO_PER_TAO == 0` for alpha deposits.

---

#### M-7: CLI event pretty-print omits TAO amounts for Deposit/Reclaimed/Slashed events

**Files**: `crates/collateral-contract/src/main.rs:579-589, 624-647, 662-715`

TAO amounts are missing from both pretty-print and JSON output. Operators cannot audit TAO flows through the CLI.

---

### LOW Severity Issues

| # | Issue | File |
|---|---|---|
| L-1 | `from_block + 1` potential overflow (use `checked_add`) | `collateral_scan.rs:44-45` |
| L-2 | `std::env::set_var` in `#[tokio::test]` is unsound in multi-threaded context | `collateral_e2e.rs:148, 209, 257` |
| L-3 | `refresh_price_cache` is a public no-op | `manager.rs:120-122` |
| L-4 | Unnecessary `node_id.clone()` in CLI query commands | `main.rs:448, 460-461` |
| L-5 | Multiple `TODO` comments -- should be tracked as issues | Multiple files |
| L-6 | Miner uses `f64` for RAO-to-alpha while validator uses `Decimal` | `basilica-miner/src/main.rs:385-390` |
| L-7 | `flow.sh` and `query.sh` have hardcoded contract addresses | `flow.sh:3`, `query.sh:3` |
| L-8 | `deploy.sh` doesn't validate env vars before use | `deploy.sh` |
| L-9 | `.env.local` written without restrictive file permissions | `setup-localnet-env.sh:360-380` |
| L-10 | `CONTRACT_DEPLOYED_BLOCK_NUMBER` is 0 -- will scan from genesis on mainnet | `config.rs:24` |
| L-11 | Private key accepted as CLI argument (visible in `/proc/*/cmdline`) | `main.rs:52` |
| L-12 | Integration test scripts lack `set -e` for error handling | `integration-test.sh` |
| L-13 | `StakingV2PrecompileMock.getStake` returns meaningless values based on gas | `CollateralUpgradeable.t.sol:21-23` |

---

## Suggestions for Improvements

1. **Add on-chain rate limiting for slashes**: A daily/weekly cap on total slashable amounts per trustee would limit blast radius of key compromise. This is the single most impactful security enhancement for the financial system.

2. **Implement block range chunking in event scanner**: Critical for first-run and catch-up scenarios. Without it, new validators cannot sync collateral state.

3. **Add idempotency checks to deposit event handler**: The reclaim handler checks for duplicates (`collateral_persistence.rs:446-456`), but deposit handler does not. If an RPC returns duplicate events, deposits could be double-counted.

4. **Consider reorg safety margin**: Re-scan from `last_block - N` with idempotent event handling to handle chain reorganizations.

5. **Share network parsing utility**: Extract to `config/collateral.rs` and fail consistently on unknown networks in both scanner and slash executor.

6. **Set `CONTRACT_DEPLOYED_BLOCK_NUMBER` to actual deployment block before mainnet**: Currently 0, which would cause full chain scan.

---

## Positive Observations

1. **CEI pattern correctly applied throughout**: All external-call-making functions properly update state before external calls. Local variables capture values before zeroing storage.
2. **ReentrancyGuard consistently on all mutative functions**: Correct use of OpenZeppelin's `ReentrancyGuardUpgradeable`.
3. **UUPS upgrade pattern properly implemented**: `_disableInitializers()` in constructor, `initializer` modifier, `reinitializer(2)` on V2, correct `_authorizeUpgrade` with role check.
4. **Atomic per-block event processing**: `apply_collateral_events_for_block` wraps all events + block number update in a single SQLite transaction. If any event fails, the entire block rolls back.
5. **Comprehensive rate limiting on slash execution**: Three-layer protection (per-miner cooldown, global rate, circuit breaker) with proper entry pruning.
6. **SHA-256 evidence integrity chain**: Slash evidence stored, hashed, checksum submitted on-chain -- creating an immutable audit trail.
7. **SQL injection prevention**: All queries use parameterized `sqlx` bindings. Column selection in `get_collateral_amount_internal` uses `match` instead of string interpolation.
8. **Saturating arithmetic throughout**: Both Solidity (OpenZeppelin Math) and Rust (U256 `saturating_add`/`saturating_sub`) prevent over/underflow.
9. **Comprehensive test suite**: 3600+ lines of Solidity tests covering deposits, reclaims, slashes, denials, access control, edge cases, upgrades. Rust-side tests cover dual-asset state machine transitions and atomic rollback.
10. **Clean separation of concerns**: Evaluator, manager, executor, scanner, evidence, and grace tracker each have single responsibilities following SOLID principles.
11. **Config validation**: `CollateralConfig::validate()` performs comprehensive parameter validation including R2 evidence config requirements when shadow mode is disabled.
12. **EOA enforcement for first deposit**: `msg.sender.code.length == 0 && tx.origin == msg.sender` prevents contract-based ownership claims. Constructor bypass also blocked.

---

## Recommendation

**Do NOT merge until C-1, C-2, C-3, C-4 are fixed.** These are state-correctness bugs in the smart contract that could lead to fund insolvency (C-1), permanent state corruption (C-2), denial of service (C-3), and operational reliability issues (C-4).

**Suggested merge path**:
1. Fix C-1 through C-4 (critical contract and Rust bugs)
2. Address H-1 through H-6 (high severity -- strongly recommended before production)
3. Create tracking issues for M-1 through M-7 (can be addressed in follow-up PRs)
4. Low issues are non-blocking

The overall architecture is sound and demonstrates strong engineering fundamentals. The dual-asset collateral model is well-designed with proper separation of TAO and alpha state. The primary concerns are around edge cases in the state machine (pending reclaim + slash interactions) and operational safety (scan loop, precision, defaults).

**Review by epappas — CHANGES_REQUESTED** [review]

File: crates/basilica-validator/migrations/017_dual_collateral_event_state.sql [hunk 1]
```diff
@@ -0,0 +1,33 @@
+-- Switch collateral persistence to dual-collateral state with reclaim request tracking.
+-- Collateral is not enabled yet, so we intentionally rebuild these tables instead of
+-- preserving legacy rows.
+
+ALTER TABLE collateral_status RENAME TO collateral_status_legacy;
+
+CREATE TABLE IF NOT EXISTS collateral_status (
```
[the remark this answers is not in the document]
Comment (itzlambda) [reply crates/basilica-validator/migrations/017_dual_collateral_event_state.sql]: its unique based on `hotkey, node_id`

File: crates/basilica-validator/migrations/017_dual_collateral_event_state.sql [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  alpha_collateral TEXT NOT NULL DEFAULT '0',
+  pending_tao_reclaim TEXT NOT NULL DEFAULT '0',
+  pending_alpha_reclaim TEXT NOT NULL DEFAULT '0',
+  url TEXT,
+  url_content_md5_checksum TEXT,
+  updated_at TEXT DEFAULT CURRENT_TIMESTAMP,
+  UNIQUE(hotkey, node_id)
+);
+
+CREATE TABLE IF NOT EXISTS collateral_reclaims (
```
[the remark this answers is not in the document]
Comment (itzlambda) [reply crates/basilica-validator/migrations/017_dual_collateral_event_state.sql]: its `reclaim_request_id` which comes from the contract and guarantees uniqueness
